### PR TITLE
Triangulation_23: Enhance IO

### LINKED ISSUE #861: I/O of Triangulation_2 or Triangulation_3, when the vertex or cell types have handles
(Cc @MoniqueTeillaud)

See the discussion in CGAL-develop:
* March 2015: https://sympa.inria.fr/sympa/arc/cgal-develop/2015-03/msg00189.html
* Followup in April 2015: https://sympa.inria.fr/sympa/arc/cgal-develop/2015-04/msg00011.html

Let's quote my mail:
> Let's take Triangulation_3 as an example. The operator<< of Triangulation_3 
> uses two maps:
>
>     // the "map" from indices to Vertex_handle
>     std::vector<Vertex_handle> TV;
>     // the map from Vertex_handle to indices
>     Unique_hash_map<Vertex_handle, std::size_t > V;
>
>The `operator<<` first displays:
> ````
>   <dimension>
>   <number of vertices>
> ````
> Then, for each vertices:
>  * it fills the `TV` and `V`
>  * and it calls the operator<< of the vertex
>
> Then:
>   * call `tds().print_cells(ostream, V);`
> 
> Then, for each cell, in dimension 3:
>   it calls the `operator<<` of the cell
> 
> Imagine that the vertex base has an extra field of type `Vertex_handle` (to 
> store for example the "parent vertex" of Steiner vertices, in Shewchuk 
> Delaunay refinement algorithm). To correctly output that triangulation, the 
> `operator<<` of the vertex should have access to the map `V`, but it does not have 
> access to it. What could be a solution for that?

### LINKED ISSUE #2359: Mesh_3: Save/load c3t3 objects correctly
## Issue Details

For the moment, the I/O of C3t3 objects only save/load the information of the TDS_3, and the I/O of `C3t3::Triangulation::Vertex` and `C3t3::Triangulation::Cell` ensures that most of the information of the C3t3 is saved and restored.

But a C3t3 object has data that is not stored in the TDS. Some of it can be retrieved after a load by a call to the *undocumented* method `c3t3.rescan_after_load_of_triangulation()`, but the `c3t3.is_in_complex(Edge)` information cannot be retrieved. It should be stored in the file, and loaded from it.

This issue will require a solution to #861 first.

## Environment

* CGAL version: CGAL-4.11

Cc: @sloriot

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Add the possibility to use functions of the cell/face/vertex base, called read/write_data(), and that have access to the maps/vectors internal to the triangulation. This is used in the IO functions.

## Release Management

* Affected package(s):Triangulation_2, Triangulation_3
* Issue(s) solved (if any): fix #861, fix #2359
* Feature/Small Feature (if any):
* Link to compiled documentation (obligatory for small feature) 
* License and copyright ownership:

### CODE REVIEW

File: Stream_support/include/CGAL/IO/has_data.h [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#ifndef CGAL_HAS_DATA_H
```
Comment (sloriot) [inline Stream_support/include/CGAL/IO/has_data.h]: No license header and `::` namespace pollution

**Comment by maxGimeno** [comment]
@sloriot @lrineau So i tried something to implement the edges in complex IO in .cgal format, but it requiered changing the signature of the Compact_mesh_cell_base_3, which changed the c3t3 type written in the header of the files. In concequences of which the demo is not able to read any of the existing "[.binary].cgal" files... I'd like thoughts on the matter.

**Comment by maxGimeno** [comment]
@sloriot @lrineau ping. 
Should I remake all existing files (in tests and examples) to fit the new type or do you have a different solution to implement the edges in complex IO that does not require changing its type ? Or something about the IO that reads the type, anything ?

**Comment by sloriot** [comment]
I'm a bit afraid that binary files are conflicting (for size reason).  Was those file updated recently?

**Comment by maxGimeno** [comment]
I did not touch this branch since april 19.

**Comment by lrineau** [comment]
> I'm a bit afraid that binary files are conflicting (for size reason). Were those files updated recently?

That is not really recently. The PR #4397 has changed the order of cells in a triangulation (given the sequence of points). That is why all the files have changed in `master`. I am surprised that this PR conflicts for the binary files, but not for the two corresponding ASCII files.

**Comment by maxGimeno** [comment]
I have not yet cupdated all the c3t3 files in this branch, as I was waiting for an aswer to my question.

**Comment by sloriot** [comment]
There is no option but to update the files if the format changes.

**Comment by maxGimeno** [comment]
Ok, I'll do that

**Comment by maxGimeno** [comment]
After working again on the branch to update it, I think that there was no conflicts on the ASCII test files because the changes were localized enough (it is just adding edges).

**Comment by lrineau** [comment]
> After working again on the branch to update it, I think that there was no conflicts on the ASCII test files because the changes were localized enough (it is just adding edges).

You can easily test with `ctest -R test_c3t3_io`.

**Comment by maxGimeno** [comment]
I know, I was just modifying my previous answer about you comment.

**Comment by maxGimeno** [comment]
Succesfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-112.shtml

**Comment by sloriot** [comment]
Tested in CGAL-5.1-Ic-114

**Comment by sloriot** [comment]
@lrineau are you done with your review?

**Comment by lrineau** [comment]
> @lrineau are you done with your review?

No, I have not reviewed the code, yet. And I will not probably do it soon.

From far, I do not really see how this PR solves #2359.

File: Mesh_3/doc/Mesh_3/CGAL/Mesh_cell_base_3.h [hunk 2]
```diff
@@ -8,7 +8,7 @@ namespace CGAL {
 Compact_mesh_cell_base_3, because the later has a different API.
 -- Laurent Rineau, 2013/10/16
 \deprecated This class is deprecated since \cgal 4.3. Use
-`CGAL::Compact_mesh_cell_base_3<Gt,MD,Tds>` instead.
+`CGAL::Compact_mesh_cell_base_3<Gt,MD,CurveIndex,Tds>` instead.
```
Comment (lrineau) [inline Mesh_3/doc/Mesh_3/CGAL/Mesh_cell_base_3.h:11]: Why add a new parameter, that is identical to `typename MD::Curve_index`?

**Comment by maxGimeno** [comment]
> > @lrineau are you done with your review?
> 
> No, I have not reviewed the code, yet. And I will not probably do it soon.
> 
> From far, I do not really see how this PR solves #2359.

Maybe I did not understand the issue then, as I thougt adding the edges in complex in the IO was the whole point.

File: Mesh_3/doc/Mesh_3/CGAL/Mesh_cell_base_3.h (hunk 2 above)
Comment (maxGimeno) [reply Mesh_3/doc/Mesh_3/CGAL/Mesh_cell_base_3.h:11]: There is no such type in the Concept MeshDomain_3, in the doc of Polyhedral_mesh_domain_3, and apparently `there is no type Curve_index in class CGAL::Polyhedral_mesh_domain_3`

Comment (lrineau) [reply Mesh_3/doc/Mesh_3/CGAL/Mesh_cell_base_3.h:11]: That is right. I understand, now.

**Comment by MaelRL** [comment]
@lrineau @maxGimeno Does this PR just get pushed to next release (that will deal with I/O of triangulations too anyway)?

**Comment by maxGimeno** [comment]
I don't see why it should be delayed to next release, it already passed the testsuite once, and it is not a small feature. I'm just waiting for @lrineau to do his review to test it again and have it integrated. Unless he explains why it does not work or fix the issue, in which case everything will have to be re thought from the start.

**Comment by sloriot** [comment]
@maxGimeno  is this replaced by #5145?

**Comment by maxGimeno** [comment]
> @maxGimeno is this replaced by #5145?

Not entirely, but if #5145 is dropped, then the second part of this PR is impossible to implement, IIRC

**Comment by lrineau** [comment]
Replaced by https://github.com/CGAL/cgal/pull/5145.

**Pull request closed without merging**